clippy_lints/src/manual_ignore_case_cmp.rs

```diff
@@ -0,0 +1,98 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet;
+use clippy_utils::ty::{is_type_diagnostic_item, is_type_lang_item};
+use rustc_errors::Applicability;
+use rustc_hir::ExprKind::{Binary, MethodCall};
+use rustc_hir::{BinOpKind, Expr, LangItem};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty;
+use rustc_middle::ty::{Ty, UintTy};
+use rustc_session::declare_lint_pass;
+use rustc_span::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for manual case-insensitive ASCII comparison.
+    ///
+    /// ### Why is this bad?
+    /// The `eq_ignore_ascii_case` method is faster because it does not allocate
+    /// memory for the new strings, and it is more readable.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// fn compare(a: &str, b: &str) -> bool {
+    ///     a.to_ascii_lowercase() == b.to_ascii_lowercase()
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// fn compare(a: &str, b: &str) -> bool {
+    ///    a.eq_ignore_ascii_case(b)
+    /// }
+    /// ```
+    #[clippy::version = "1.82.0"]
+    pub MANUAL_IGNORE_CASE_CMP,
+    nursery,
+    "manual case-insensitive ASCII comparison"
+}
+
+declare_lint_pass!(ManualIgnoreCaseCmp => [MANUAL_IGNORE_CASE_CMP]);
+
+fn get_ascii_type<'tcx>(cx: &LateContext<'tcx>, expr: &Expr<'_>) -> Option<Ty<'tcx>> {
+    let ty_raw = cx.typeck_results().expr_ty(expr);
+    let ty = ty_raw.peel_refs();
+    if needs_ref_to_cmp(cx, ty)
+        || ty.is_str()
+        || ty.is_slice()
+        // FIXME: Need a better way to check if ty is OsStr(ing)
+        || ["std::ffi::OsStr", "std::ffi::OsString"].contains(&ty.to_string().as_str())
```

Comment:
You can use the diagnostic items `OsStr` and `OsString`. The code would look roughly, like this:

```rs
|| matches!(
        get_type_diagnostic_name(cx, ty),
        Some(sym::OsStr | sym::OsString)
    )
```

Or you can also add two checks with `is_type_diagnostic_item(cx, ty, sym::OsStr)`...